tests/azure_iot_provisioning_client_ut.c

```diff
@@ -0,0 +1,704 @@
+/* Copyright (c) Microsoft Corporation. All rights reserved. */
+/* SPDX-License-Identifier: MIT */
+
+#include <stdarg.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <setjmp.h>
+
+#include <cmocka.h>
+
+#include "azure_iot_mqtt.h"
+#include "azure_iot_provisioning_client.h"
+/*-----------------------------------------------------------*/
+
+/* Data exported by cmocka port for MQTT */
+extern AzureIoTMQTTPacketInfo_t xPacketInfo;
+extern AzureIoTMQTTDeserializedInfo_t xDeserializedInfo;
+extern uint16_t usTestPacketId;
+extern const uint8_t * pucPublishPayload;
+/*-----------------------------------------------------------*/
+
+static const uint8_t ucEndpoint[] = "unittest.azure-devices-provisioning.net";
+static const uint8_t ucIdScope[] = "0ne000A247E";
+static const uint8_t ucRegistrationId[] = "UnitTest";
+static const uint8_t ucSymmetricKey[] = "ABC12345";
+static const uint8_t ucTestProvisioningServiceResponse[] = "$dps/registrations/res/202/?";
+static uint8_t ucBuffer[ 1024 ];
+static AzureIoTTransportInterface_t xTransportInterface =
+{
+    .pxNetworkContext = NULL,
+    .xSend            = ( AzureIoTTransportSend_t ) 0xA5A5A5A5,
+    .xRecv            = ( AzureIoTTransportRecv_t ) 0xACACACAC
+};
+static const uint8_t ucAssignedHubResponse[] = "{ \
+    \"operationId\":\"4.002305f54fc89692.b1f11200-8776-4b5d-867b-dc21c4b59c12\",\"status\":\"assigned\",\"registrationState\": \
+         {\"registrationId\":\"reg_id\",\"createdDateTimeUtc\":\"2019-12-27T19:51:41.6630592Z\",\"assignedHub\":\"test.azure-iothub.com\", \
+          \"deviceId\":\"testId\",\"status\":\"assigned\",\"substatus\":\"initialAssignment\",\"lastUpdatedDateTimeUtc\":\"2019-12-27T19:51:41.8579703Z\", \
+          \"etag\":\"XXXXXXXXXXX=\"\
+         }\
+}";
+static const uint8_t ucAssigningHubResponse[] = "{ \
+    \"operationId\":\"4.002305f54fc89692.b1f11200-8776-4b5d-867b-dc21c4b59c12\",\"status\":\"assigning\" \
+}";
+static const uint8_t ucFailureHubResponse[] = "{ \
```

Comment:
Could we add additional bad responses in JSON blobs - from horrible stuff like not even JSON in 1st place to something that's close to right JSON but that maybe missing field or two that either embedded SDK or potentially middleware layer would be unhappy on missing?

Reply:
parsing is done by embedded sdk, and they already have these tests.. Just wondering if this will add value. But in the next PR to increase the line coverage I will add additional tests.

---
In reply to: [608075635](https://github.com/Azure/azure-iot-middleware-freertos/pull/62#discussion_r608075635) [](ancestors = 608075635)

Reply:
I think we should add at least one test to make the embedded SDK fail.  This will test that if we get bad JSON & embedded fails, then our layer also properly handles the scenario.

Reply:
Sure, I will add in next PR.